src/mount.cpp

```diff
@@ -0,0 +1,164 @@
+{
+  char* p  = realpath(path.c_str(), nullptr);
+  if (p) {
+    std::string ret(p);
+    free(p);
+    return ret;
+  } else {
+    char* err = strerror(errno);
+    return util::unexpected(err);
+  }
+}
+
+bool is_valid_image_and_mountpoint(const std::string &mount_point,
+                              const std::string &squashfs_file) {
+  struct stat mnt_stat;
+  auto mnt_status = stat(mount_point.c_str(), &mnt_stat);
+  if (mnt_status) {
+    slurm_spank_log("Invalid mount point \"%s\"", mount_point.c_str());
+    return false;
+  }
+  if (!S_ISDIR(mnt_stat.st_mode)) {
+    slurm_spank_log("Invalid mount point \"%s\" is not a directory",
+                    mount_point.c_str());
+    return false;
+  }
+  // Check that the input squashfs file exists.
+  int sqsh_status = stat(squashfs_file.c_str(), &mnt_stat);
+  if (sqsh_status) {
+    slurm_spank_log("Invalid squashfs image \"%s\"", squashfs_file.c_str());
+    return false;
+  }
+  if (!S_ISREG(mnt_stat.st_mode)) {
+    slurm_spank_log("Invalid squashfs image \"%s\" is not a file",
+                    squashfs_file.c_str());
+    return false;
+  }
+
+  return true;
+}
+
+int do_mount(spank_t spank, const std::vector<mount_entry>& mount_entries) {
+  if(mount_entries.size() == 0)
+    return ESPANK_SUCCESS;
+  if (unshare(CLONE_NEWNS) != 0) {
+    slurm_spank_log("Failed to unshare the mount namespace");
+    return -ESPANK_ERROR;
+  }
+  // make all mounts in new namespace slave mounts, changes in the original
+  // namesapce won't propagate into current namespace
+  if (mount(NULL, "/", NULL, MS_SLAVE | MS_REC, NULL) != 0) {
+    slurm_spank_log("mount: unable to change `/` to MS_SLAVE | MS_REC");
+    return -ESPANK_ERROR;
+  }
+
+  for (auto& entry : mount_entries){
+    if (entry.p != protocol::file) {
+      slurm_spank_log("Image must be given as local file (jfrog://, https:// not implemented)");
+      return -ESPANK_ERROR;
+    }
+    std::string mount_point = entry.mount_point;
+    std::string squashfs_file = entry.image_path;
+    // Check that the mount point exists.
+    struct stat mnt_stat;
+    auto mnt_status = stat(mount_point.c_str(), &mnt_stat);
+    if (mnt_status) {
+      slurm_spank_log("Invalid mount point \"%s\"", mount_point.c_str());
+      return -ESPANK_ERROR;
+    }
+    if (!S_ISDIR(mnt_stat.st_mode)) {
+      slurm_spank_log("Invalid mount point \"%s\" is not a directory",
+                      mount_point.c_str());
+      return -ESPANK_ERROR;
+    }
+
+    // Check that the input squashfs file exists.
+    int sqsh_status = stat(squashfs_file.c_str(), &mnt_stat);
+    if (sqsh_status) {
+      slurm_spank_log("Invalid squashfs image \"%s\"", squashfs_file.c_str());
+      return -ESPANK_ERROR;
+    }
+    if (!S_ISREG(mnt_stat.st_mode)) {
+      slurm_spank_log("Invalid squashfs image \"%s\" is not a file",
+                      squashfs_file.c_str());
+      return -ESPANK_ERROR;
+    }
```

Comment:
We can use is_valid_image_and_mountpoint() in init_post_opt_remote() and init_post_opt_local_allocator(). The routine is called before do_mount(). That way, do_mount can assume all mount points are valid and images exist as a (documented) prerequisite. This also ensures error and log messages are consistent between local and remote contexts.

Reply:
`do_mount` does also verify the image path and mountpoint is valid. The routine `is_valid_image_and_mountpoint()` is run in allocator and local context to ensure that the exit status is `>0` in case there is an error. Iirc if the failure happens only in remote context, the job fails, but the exit status is zero.

Reply:
I was suggesting to remove the code that checks validity in do_mount() and reuse the code in is_valid_image_and_mountpoint().